### ISSUE TITLE
put membership in the JWT or on the server socket

### ISSUE THREAD

**Issue by mattkrick** [ParabolInc/action#167]
Right now, for every read and mutation, we have to first look up the userId in the database to make sure that they belong to the team.
This is gonna become a bottleneck as we approach realtime, because doing something as simple as changing `foo` to `food` will require 2 DB hits.

We can stick an array of teamMemberIds in the JWT.
Or, we can stick it on the server socket.

As we socketize more of the app, a larger JWT isn't a terrible thing, since it's stored on the server socket anyways. 

The only question left is what to do for HTTP requests, where the question simply becomes: "is a larger payload worth an extra DB lookup"? Chances are, yes.

**Comment by jordanh** [comment]
Chances are, yes. I agree.

**Comment by mattkrick** [comment]
- The hottest mutation in our app is going to be `updateTask`.
- This mutation requires making sure that the user calling the mutation is on the team that the task belongs to
- Currently, we get the `teamId` of the task by referencing the `teamMemberId` of the task assignee
- Then, we get all the teamMembers with that teamId, and filter by the authToken's userId

Risking premature optimization, this seems really expensive for something that we will be calling on every keystroke (or every throttled keystroke).
Some options:
- Put a `teamId` index on the Task. Not a huge fan of this since we'll be writing to it more than reading & we've already got a couple indexes on this table
- design `teamMemberId` so that it is a `teamId` + incrementing number (eg `team123::1` is the teamMemberId for someone on `team123`)
- put an array of `teamMemberId` inside the JWT.

The last 2 aren't mutually exclusive. If we implemented both, then we could get the `teamId` from the task without a DB hit and we could get the list of teams the user is on from the authToken. That means we could accomplish this whole thing without a single hit to the DB.

**Comment by mattkrick** [comment]
The only security concern is whether it is OK that our teamMemberIds are predictable, given a teamId. I don't _think_ that's a problem, since you can't do much by just knowing the teamMember's id. We could solve this by using another shortid `team123::member456` if really necessary.

**Comment by jordanh** [comment]
I prefer option #1 here. I mean, disk-space is cheap. If we can solve this with an index, go for it. Since teamId won't be changing, the write performance should stay high.

**Comment by mattkrick** [comment]
ah, so adding a teamId index doesn't get us all the way there.

here's a quick landscape of our options:

```
//current: 4 hits
const {teamMemberId} = await r.table('Task').get(updatedTask.id);
const {teamId} = await r.table('TeamMember').get(teamMemberId);
const canwrite = await r.table('TeamMember').getAll(teamId, {index: 'teamId'}).filter({userId: authToken.sub}).isEmpty().not();
if (canwrite) {
  await r.table('Task').update(updatedTask);
}

// add teamId to the Task (option 1): 3 hits
const {teamId} = await r.table('Task').get(updatedTask.id);
const canwrite = await r.table('TeamMember').getAll(teamId, {index: 'teamId'}).filter({userId: authToken.sub}).isEmpty().not();
if (canwrite) {
  await r.table('Task').update(updatedTask);
}

// add teams to JWT (option 2): 3 hits (but a little cheaper than option 1)
const {teamMemberId} = await r.table('Task').get(updatedTask.id);
const {teamId} = await r.table('TeamMember').get(teamMemberId);
const canwrite = authToken.teams.includes(teamId);
if (canwrite) {
  await r.table('Task').update(updatedTask);
}

// make teamMemberId = userId::teamId (option 3): 3 hits (equal to option 1)
const {teamMemberId} = await r.table('Task').get(updatedTask.id);
const [userId, teamId] = teamMemberId.split('::');
const canwrite = await r.table('TeamMember').getAll(teamId, {index: 'teamId'}).filter({userId: authToken.sub}).isEmpty().not();
if (canwrite) {
  await r.table('Task').update(updatedTask);
}

// options 2 & 3: 2 hits
const {teamMemberId} = await r.table('Task').get(updatedTask.id);
const [userId, teamId] = teamMemberId.split('::');
const canwrite = authToken.teams.includes(teamId);
if (canwrite) {
  await r.table('Task').update(updatedTask);
}
```

**Comment by mattkrick** [comment]
and a bonus that results in a single DB hit:

```
// make taskId = teamId::taskId (option 2 + bonus)
const [teamId, taskId] = updatedTask.id.split('::');
const canwrite = authToken.teams.includes(teamId);
if (canwrite) {
  await r.table('Task').update(updatedTask);
}
```

**Comment by mattkrick** [comment]
fixed by #181
